File: ch6/str_align.cpp
#include <algorithm>
#include <cstdio>
#include <cstring>
using namespace std;

int main() {
  char A[20] = "ACAATCC", B[20] = "AGCATGC";
  int n = (int)strlen(A), m = (int)strlen(B);
  int i, j, table[20][20]; // Needleman Wunsnch's algorithm

  memset(table, 0, sizeof table);
  // insert/delete = -1 point
  for (i = 1; i <= n; i++)
    table[i][0] = i * -1;
  for (j = 1; j <= m; j++)
    table[0][j] = j * -1;

  for (i = 1; i <= n; i++)
    for (j = 1; j <= m; j++) {
      // match = 2 points, mismatch = -1 point
      table[i][j] = table[i - 1][j - 1] + (A[i - 1] == B[j - 1] ? 2 : -1); // cost for match or mismatches
      // insert/delete = -1 point
      table[i][j] = max(table[i][j], table[i - 1][j] - 1); // delete
      table[i][j] = max(table[i][j], table[i][j - 1] - 1); // insert
    }

  printf("DP table:\n");
  for (i = 0; i <= n; i++) {
    for (j = 0; j <= m; j++)
      printf("%3d", table[i][j]);
    printf("\n");
  }
  printf("Maximum Alignment Score: %d\n", table[n][m]);

  return 0;
}

File: ch6/str_align.py
def main():
    A = "ACAATCC"
    B = "AGCATGC"
    n = len(A)
    m = len(B)

    table = [[0 for j in range(20)] for i in range(20)] # Needleman Wunsnch's algorithm

    for i in range(1, n+1):
        table[i][0] = i * -1
    for j in range(1, m+1):
        table[0][j] = j * -1

    for i in range(1, n+1):
        for j in range(1, m+1):
            # match = 2 points, mismatch = -1 point
            table[i][j] = table[i - 1][j - 1] + (2 if A[i - 1] == B[j - 1] else -1)
            # insert/delete = -1 point
            table[i][j] = max(table[i][j], table[i - 1][j] - 1); # delete
            table[i][j] = max(table[i][j], table[i][j - 1] - 1); # insert

    print("DP table:")
    for i in range(0, n+1):
        for j in range(0, m+1):
            print("{:>3}".format(table[i][j]), end='')
        print()

    print("Maximum Alignment Score: {}".format(table[n][m]))

main()
